numpy/core/src/multiarray/einsum_sumprod.c.src

```diff
@@ -1542,6 +1542,113 @@ bool_sum_of_products_outstride0_@noplabel@(int nop, char **dataptr,
 #  endif
 }
 
+/* For now, a catch all function for objects */
+static void
+object_sum_of_products_@noplabel@(int nop, char **dataptr,
+                                npy_intp const *strides, npy_intp count)
+{
+    PyObject *tmp1, *tmp2 = NULL;
+
+    while(count--){
+        tmp1 = *(PyObject **)dataptr[0];
+        if(!tmp1){
+            return;
+        }
+        int i;
+        for(i = 1; i < nop; ++i){
+            Py_XDECREF(tmp2);
+            if((*(PyObject **)dataptr[i]) == NULL){
+                return;
+            }
+            tmp2 = PyNumber_Multiply(tmp1, *(PyObject **)dataptr[i]);
+            if(!tmp2){
+                return;
+            }
+            tmp1 = tmp2;
+        }
+
+        if(*(PyObject **)dataptr[i] == NULL){
+            Py_XDECREF(tmp2);
+            return;
+        }
+
+        tmp2 = PyNumber_Add(tmp1, *(PyObject **)dataptr[i]);
+        if(i != 0){
+            Py_XDECREF(tmp1);
+        }
+        if(!tmp2){
+            return;
+        }
+        Py_XDECREF(*(PyObject **)dataptr[nop]);
+        *(PyObject **)dataptr[nop] = tmp2;
+        tmp2 = 0;
+        for (i = 0; i <= nop; ++i){
+            dataptr[i] += strides[i];
+        }
+    }
+}
+/**end repeat**/
+
+/**begin repeat
+ * #fn_name = object_sum_of_products_contig_one,
+ *  object_sum_of_products_contig_two, 
+ *  object_sum_of_products_stride0_contig_outcontig_two,
+ *  object_sum_of_products_contig_stride0_outcontig_two,
+ *  object_sum_of_products_stride0_contig_outstride0_two,
+ *  object_sum_of_products_contig_stride0_outstride0_two,
+ *  object_sum_of_products_contig_contig_outstride0_two,
+ *  object_sum_of_products_contig_three,
+ *  object_sum_of_products_contig_any,
+ *  object_sum_of_products_contig_outstride0_one,
+ *  object_sum_of_products_outstride0_one,
+ *  object_sum_of_products_outstride0_two,
+ *  object_sum_of_products_outstride0_three,
+ *  object_sum_of_products_outstride0_any#
+ */
+static void
+@fn_name@(int nop, char **dataptr,
+                                npy_intp const *strides, npy_intp count)
+{
+    PyObject *tmp1, *tmp2 = NULL;
+
+    while(count--){
+        tmp1 = *(PyObject **)dataptr[0];
+        if(!tmp1){
+            return;
+        }
+        int i;
+        for(i = 1; i < nop; ++i){
+            Py_XDECREF(tmp2);
+            if((*(PyObject **)dataptr[i]) == NULL){
+                return;
+            }
+            tmp2 = PyNumber_Multiply(tmp1, *(PyObject **)dataptr[i]);
+            if(!tmp2){
+                return;
```

Comment:
This code is reachable if `__mul__` raises. Does it do the right thing already in that case, or does it raise SystemError?

Reply:
Hmm. Should there be a PyErr_Clear() in that if statement?

Reply:
If an exception is thrown but `einsum` returns a non-`NULL` object, then python will throw `SystemError`. You have to make sure that if an exception gets set, it gets propagated.

Reply:
I'm looking at [the documentation](https://docs.python.org/3.9/extending/extending.html#intermezzo-errors-and-exceptions) to propogate an exception, it says to return NULL, like you said. However, the other functions in _einsum_sumprod.c.src_ seem to all be void type. And when the function is called in _einsum.c_  like:
`sop(2, ptrs[0], strides[0], shape[0]);`
there is no check for a failure. Should this behavior be changed?

Reply:
Right, this is exactly what I was alluding to. Your best options here are either:
* change `sop` to return an `int`, and return `-1` on failure, `0` on success. You can then propagate that as appropriate
* leave it as `void`, but use `PyErr_Occurred()` to check if an exception happened in `sop`. This is generally the workaround we use when there's no channel to pass an error bit back through

Note that the first approach is likely to conflict with @Qiyu8 / @seiko2plus' work to speed up einsum.

Reply:
Ohh... and now I'm thinking the newly created objects in _data_out_ need to be `Py_xDECREF`'d in _einsum.c_ if an error occured

Reply:
That depends if `data_out` is being managed by an array somewhere, in which case the cleanup handler for `ndarray` will decref them for you. If it's just a raw C array, then your conclusion is correct.

Reply:
`data_out` or `dataptr[nop]` in this function, comes from `NpyIter` and is related to `PyArrayObject *op[NPY_MAXARGS]` but my understanding of this relationship ends there. So I guess I will do some research over the holidays to figure this out if the PyArrayObject in `op[nop]`  owns the newly created objects or if something else/no one else does.

Reply:
If the data comes from a `PyArrayObject` then you don't have to worry about the reference counts, it will xdecref all its elements when the array itself is decref'd - even if you filled only half.

There is probably a special flag for reference-counted items that `NpyIter` uses, but I don't remember it right now.